vllm/compilation/backends.py

```diff
@@ -164,106 +166,285 @@ def wrap_inductor(graph, example_inputs, additional_inductor_config):
+        self.is_first_graph = is_first_graph
+
+        self.compile_sizes: Set[int] = set(
+            self.compilation_configs.compile_sizes)
+        self.capture_sizes: Set[int] = set(
+            self.compilation_configs.capture_sizes)
+
+        self.compile_finished = False
+        self.first_run_finished = False
+
+        self.compiled_graph_for_general_shape: Callable = None  # type: ignore
+
+        self.sym_shape_indices: List[int] = []
+
+        # the entries for different shapes that we need to either
+        # compile or capture cudagraph
+        self.concrete_size_entries: Dict[int, ConcreteSizeEntry] = {}
+        for shape in self.compile_sizes.union(self.capture_sizes):
+            self.concrete_size_entries[shape] = ConcreteSizeEntry(
+                runtime_shape=shape,
+                need_to_compile=shape in self.compile_sizes,
+                use_cudagraph=shape in self.capture_sizes,
+            )
+
+    def __call__(self, *args) -> Any:
+
+        if not self.compile_finished:
+            self.compile_finished = True
+
+            # this is the first compilation, we will compile a graph with
+            # dynamic shape, as the caller will mark first dimension as dynamic
+
+            self.sym_shape_indices = [
+                i for i, x in enumerate(args) if isinstance(x, torch.SymInt)
+            ]
+
+            if self.compilation_configs.use_inductor:
+                # args here are example inputs (fake tensors)
+                if self.is_first_graph:
+                    logger.info("Compiling a graph for general shape")
+                self.compiled_graph_for_general_shape = wrap_inductor(
+                    self.graph, args,
+                    self.compilation_configs.inductor_compile_config)
+            else:
+                self.compiled_graph_for_general_shape = self.graph
+
+            return self.graph(*args)
+
+        if not self.first_run_finished:
+            self.first_run_finished = True
+            return self.compiled_graph_for_general_shape(*args)
+
+        runtime_shape = args[self.sym_shape_indices[0]]
+        if runtime_shape not in self.concrete_size_entries:
+            # we don't need to do anything for this shape
+            return self.compiled_graph_for_general_shape(*args)
+
+        entry = self.concrete_size_entries[runtime_shape]
+
+        if entry.runnable is None:
+            entry.runnable = self.compiled_graph_for_general_shape
+
+        if entry.need_to_compile and not entry.compiled:
+            entry.compiled = True
+            # args are real arguments
+            if self.compilation_configs.use_inductor:
+                if self.is_first_graph:
+                    logger.info("Compiling a graph for shape %s",
+                                runtime_shape)
+                entry.runnable = wrap_inductor(
+                    self.graph, args,
+                    self.compilation_configs.inductor_compile_config)
```

Comment:
Nit: this could be a method as it happens in multiplle places, perhaps `compile_for_shape`:
```python
def compile_for_shape(...):
    if not self.compilation_configs.use_inductor:
        return graph
    return wrap_inductor(...)
```

Reply:
Actually, is there any benefit to having the entries if `self.compilation_configs.use_inductor` is `False`? We could perhaps just assert `use_inductor` if any compile shapes are given.

Reply:
it's because cudagraph needs to be compiled first, if we need to compile it.

Reply:
> this could be a method as it happens in multiplle places

good point, added directly in `wrap_inductor` in [a9ae1ae](https://github.com/vllm-project/vllm/pull/9715/commits/a9ae1aec2c812795bb4d1b76bb93ed4c0d1c8bc5)